### DISCUSSION TITLE
Pagination questions...

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#38417] [Help]
So I have an old (not so old) lazy pagination code that I need to fix/change ...  eventually one day I need to scrap it and re-do it ...    BUT for now... if I can have a quick fix that would help...

I know this is not related to Next.js... or it might be....  but any inputs would help...

QUESTIONS:

1. Is there any sources to learn how to do proper pagination?   ... of course this is something more complex and I need to put on some time to study. 

2. Could anybody point out why is this pagination repeating the pagination items over again?  <-- this will be the quick fix just to make this not so great looking pagination kind of work.

one complete code sample is here:
https://github.com/Elindo586/nextprojectb/blob/main/pages/invent/charlynn.js

Here is a sample of the problem... for some pages it works fine and it doesn't repeat itself, for other  pages the pagination seems to work OK

![363636](https://user-images.githubusercontent.com/93617433/177806970-25b2311c-0ff2-40f7-bae5-9d0590af8ec4.PNG)

**Comment by Elindo586** [comment]
Eh.....

Apparently it works fine after build and checked in the browser....


I am still going to change that ugly pagination in the future...  any references to learn how to do it nicely is welcomed!

**Comment by icyJoseph** [comment]
Hi, 

Good that it works! I had started to read the code and one thing was that the patterns used are not very common, so it took a bit longer than expected.

I believe that in dev mode React Strict Mode is causing issues, or rather giving you a hint, because the strict mode renders things twice, which means the effect is running two times, https://reactjs.org/docs/strict-mode.html#ensuring-reusable-state. If this goes over your head, just know that, this is triggered on purpose to help you figure out where, _future features_ might break your code.

I think any guide you find for pagination in React will use patterns that you can't really apply to the codebase as it stands. The fact that it double renders is a helping hand from React, to let you know that things are not really in good shape.

I'll still take a look and try to come up with pointers on how to refactor the page. 

I looked at this article, https://www.freecodecamp.org/news/build-a-custom-pagination-component-in-react/, and it seemed like a good resource on pagination. 

In a broad sense, pagination is about chunking data, and through a pointer, select one of the chunks. Visually you render two things, the current chunk, and a list of all possible pointers. For accessibility, we usually show which pointer is currently selected. Here's a more framework agnostic approach to pagination https://www.a11ymatters.com/pattern/pagination/. 

I'll go through my links to see if I can find something else.
